File: simulateur_prof/DOA_Data_DeepLearning/Library/RadarChannel.m
%% Radar Free Space Channel
function y = RadarChannel(x,fc,fs,R,TwoWay,v)

% x,fc,fs,tgtpos,TwoWay,tgtvel
%%
if nargin < 5
    TwoWay = 1;
    v = 0;
elseif nargin < 6
    v = 0;
end

%%
Ts = 1/fs;  
c = physconst('lightspeed');        % Vitesse de la lumière (m/s)
lambda = c/fc;                      % longueur d'onde du radar (m)

%%
if size(R,2) == 2
    xd = R(:,1); yd = R(:,2);
    R = sqrt(xd.^2 + yd.^2);
    an = atan(yd/xd);
    an = cos(an);
else
    R = R(:,1);
    an = 1;
end

%%
if TwoWay == 1
    t = 2*R/c;                          % Retard de propagation aller-retour (s)
    fsloss = (lambda.^2./(4*pi*R).^2);     % trajet aller
    fsloss = fsloss.^2;                       % trajet retour
else
    t = R/c;                          % Retard de propagation aller-retour (s)
    fsloss = (lambda.^2./(4*pi*R).^2);     % trajet aller
end



%%
Ndelay = t/Ts;
N = round(Ndelay);
fracDelay = Ndelay-N;

% VariableFractionalDelay

%%
w = sqrt(fsloss).*exp(-1i*2*pi*2*R/lambda);  % Amplitude et phase dû à la propagation

%%
y = zeros(length(N),length(x));

%%
if all(N < length(x))
    for i = 1:length(N)
        if length(N) == size(x,1)
            y(i,:) = [zeros(1,N(i)) w(i)*x(i,1:end-N(i))];
        elseif size(x,1) == 1
            y(i,:) = [zeros(1,N(i)) w(i)*x(1,1:end-N(i))];
        else
            error('error !')
        end
    end
    if nargin == 6
        v = v*an;
        if TwoWay == 1
            fd = 2*v/lambda;
        else
            fd = v/lambda;
        end
        mfd = fd*(0:length(y)-1);
        Av = exp(-1i*2*pi*(mfd*Ts));
        y = Av.*y;
    end
else
    error(['error:  Rmax = ' num2str(round(length(x)*Ts*c/2*1e-3,2)) ' Km'])
end

%%

% fd = 2*v/lambda;
% m = 0:length(x)-1;
% Av = exp(1i*2*pi*fd*m*Ts);
end

File: simulateur_prof/Simulateur_python/RadarChannel.py
from simulateur_prof.Simulateur_python.sim_utils import get_mode_calcul

if get_mode_calcul() == "gpu":
    import cupy as np
else:
    import numpy as np
from scipy.constants import c


def radar_channel(x, fc, fs, R, two_way=True, v=0):
    """
    Simule le canal radar en espace libre.

    Parameters:
    - x: Signal d'entrée
    - fc: Fréquence centrale (Hz)
    - fs: Fréquence d'échantillonnage (Hz)
    - R: Distance entre le radar et la cible (peut être une matrice [x, y])
    - two_way: Indique s'il s'agit d'un aller-retour (True) ou seulement un aller simple (False)
    - v: Vitesse de la cible (m/s)

    Returns:
    - y: Signal de sortie après le canal radar
    """

    Ts = 1 / fs
    wavelength = c / fc

    if R.shape[1] == 2:
        xd, yd = R[:, 0], R[:, 1]
        R = np.sqrt(xd**2 + yd**2)
        an = np.cos(np.arctan2(yd, xd))
    else:
        R = R[:, 0]
        an = 1

    if two_way:
        t = 2 * R / c
        fs_loss = wavelength**2 / (4 * np.pi * R) ** 2
        fs_loss = fs_loss**2
    else:
        t = R / c
        fs_loss = wavelength**2 / (4 * np.pi * R) ** 2

    Ndelay = t / Ts
    N = np.round(Ndelay)
    frac_delay = Ndelay - N

    w = np.sqrt(fs_loss) * np.exp(-1j * 2 * np.pi * 2 * R / wavelength)

    y = np.zeros((len(N), x.shape[1]), dtype=np.complex128)

    if np.all(N < x.shape[1]):
        for i in range(len(N)):
            if len(N) == x.shape[0]:
                y[i, :] = np.concatenate(
                    [np.zeros(int(N[i])), w[i] * x[i, : int(-N[i])]]
                )
            elif x.shape[0] == 1:
                y[i, :] = np.concatenate(
                    [np.zeros(int(N[i])), w[i] * x[0, : int(-N[i])]]
                )
            else:
                raise ValueError("Error!")

        if v != 0:
            v = v * an
            if two_way:
                fd = 2 * v / wavelength
            else:
                fd = v / wavelength
            mfd = fd * np.arange(len(y))
            Av = np.exp(-1j * 2 * np.pi * mfd * Ts)
            y = Av * y
    else:
        raise ValueError(f"Error: Rmax = {round(len(x) * Ts * c / 2 * 1e-3, 2)} Km")

    return y
